### PR TITLE
Create material points within elements fixes #14, #18, #19

### LINKED ISSUE #14: Distinguish `vertices` from `points`
* Vertices from meshes should be stored separately from material points. There should be a clear distinction between these.
Depends on #1 and #3

### PR DESCRIPTION
**Pull request by cw646**
New pull request, removes Eigen files.

### CODE REVIEW

File: include/gmsh.h [hunk 1]
```diff
@@ -1,15 +1,18 @@
 #ifndef MPM_POINT_GEN_GMSH_H_
 #define MPM_POINT_GEN_GMSH_H_
 
+#include "mesh.h"
```
Comment (kks32) [inline include/gmsh.h]: These local include files should be at the end after `Eigen` includes

File: include/gmsh.h [hunk 2]
```diff
@@ -21,17 +24,28 @@ class GMSH : public Mesh {
...
+    return materialpoints_;
+  }
+
+  //! Return a map of mesh element vertices
+  std::map<double, std::array<double, 3>> vertices() const { return vertices_; }
 
-  //! Return a vector of elements
-  std::vector<std::shared_ptr<Point<3>>> elements() const { return elements_; }
+  //! Return a map of element id & vertices id
+  std::map<double, std::array<double, 4>> elements() const { return elements_; }
```
Comment (kks32) [inline include/gmsh.h]: IDs should be unsigned or int.

File: include/gmsh.h [hunk 3]
```diff
@@ -1,15 +1,18 @@
...
 #include <memory>
 #include <vector>
 
-#include "mesh.h"
-#include "point.h"
+// libraries
+#include <Eigen/Dense>
 
 //! \brief Generate Material Points from GMSH file
 class GMSH : public Mesh {
```
Comment (kks32) [inline include/gmsh.h]: Make class templatise with dimension. This avoids defining dimension elsewhere within GMSH

File: include/gmsh.h (hunk 2 above)
Comment (kks32) [inline include/gmsh.h]: Is the number `4` fixed for all elements?

File: include/gmsh.h [hunk 4]
```diff
@@ -40,15 +54,21 @@ class GMSH : public Mesh {
   //! Number of vertices
   unsigned nvertices_;
 
-  //! Vector of vertices 
-  std::vector<std::shared_ptr<Point<3>>> vertices_;
-
-  //! Vector of elements
-  std::vector<std::shared_ptr<Point<3>>> elements_;
+  std::vector<std::shared_ptr<Point<3>>> materialpoints_;
```
Comment (kks32) [inline include/gmsh.h]: This should be from the template argument of dimension `Tdim`

File: include/gmsh.h [hunk 5]
```diff
@@ -40,15 +54,21 @@ class GMSH : public Mesh {
...
+  //! $\tau_{xy}$
   std::vector<std::array<double, 6>> stress_;
+
+  //! Map to store id and vertices coordinates
+  std::map<double, std::array<double, 3>> vertices_;
+
+  //! Map to store element ID and vertices ID
+  // 4 = element vertices
+  // 12 = total number of coordinate values for vertices
+  std::map<double, std::array<double, 4>> elements_;
```
Comment (kks32) [inline include/gmsh.h]: Change it to a map of unsigned to a vector of unsigned.

File: include/gmsh.h [hunk 6]
```diff
@@ -40,15 +54,21 @@ class GMSH : public Mesh {
...
+  std::map<double, std::array<double, 12>> elementcoordinates_;
```
Comment (kks32) [inline include/gmsh.h]: These coordinates are vertices coords, right?

File: include/gmsh.tcc [hunk 7]
```diff
@@ -66,9 +64,13 @@ void GMSH::read_elements(const std::string& filename) {
 
   const unsigned toplines = 4;
   const unsigned ndimension = 3;
```
Comment (kks32) [inline include/gmsh.tcc]: Use the template dimnsion.

File: include/gmsh.tcc [hunk 8]
```diff
@@ -66,9 +64,13 @@ void GMSH::read_elements(const std::string& filename) {
...
+  // specify element type 4 = tetrahedral
+  const unsigned element_type = 4;
```
Comment (kks32) [inline include/gmsh.tcc]: Use an ENUM

File: include/gmsh.tcc [hunk 9]
```diff
@@ -109,23 +111,107 @@ void GMSH::read_elements(const std::string& filename) {
...
+  const unsigned lastelement = elements_.rbegin()->first;
+
+  std::map<double, std::array<double, 4>>::iterator elementfind;
+  std::map<double, std::array<double, 3>>::iterator verticesfind;
+
+  std::array<double, n_vertices> elementkeyvalues;
+  std::array<double, ncoordinates> verticesarray;
+
+  //! Iterate through element_
+  for (unsigned q = firstelement; q <= lastelement; ++q) {
```
Comment (kks32) [inline include/gmsh.tcc]: I thought ezra's latest commit fixed this algorithm

Comment (kks32) [reply include/gmsh.tcc]:
```
+      //! Iterate through the 4 vertices to get coordinates
+      for (unsigned q = 0; q <= 3; ++q) {
+        //! Get the vertex wanted from the id 
+        verticesfind = vertices_.find(elementkeyvalues[q]);
+        //! For each vertex, store the coordinates
+        //! j = 0 -> [X], j = 0 -> [Y], j = 0 -> [Z]
+        for (unsigned j = 0; j <= 2; ++j) {
+          verticesarray[q * 3 + j] = verticesfind->second[j];
         }
       }
```

File: include/gmsh.tcc [hunk 10]
```diff
@@ -109,23 +111,107 @@ void GMSH::read_elements(const std::string& filename) {
...
+      //! Store coordinates in 3x4 matrix
+      Eigen::MatrixXd m(3, 4);
+      for (unsigned i = 0; i <= 3; ++i) {
+        for (unsigned j = 0; j <= 2; ++j) {
+          arrayposition = (i * 3) + j;
+          m(j, i) = coordinatesfind->second[arrayposition];
+        }
+      }
+      // Centroid test
+      //++++++++++++++++++++++++++++++++++++++
```
Comment (kks32) [inline include/gmsh.tcc]: Can you iterate here as well?

File: include/gmsh.tcc [hunk 11]
```diff
@@ -109,23 +111,107 @@ void GMSH::read_elements(const std::string& filename) {
...
+
+  const unsigned firstelementcoord = elementcoordinates_.begin()->first;
+  const unsigned lastelementcoord = elementcoordinates_.rbegin()->first;
+
+  for (unsigned t = firstelementcoord; t < lastelementcoord + 1; ++t) {
+    coordinatesfind = elementcoordinates_.find(t);
+    if (coordinatesfind != elementcoordinates_.end()) {
+      //! Store coordinates in 3x4 matrix
+      Eigen::MatrixXd m(3, 4);
+      for (unsigned i = 0; i <= 3; ++i) {
```
Comment (kks32) [inline include/gmsh.tcc]: Why do they have different indices 3 and 2?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/gmsh.h [hunk 12]
```diff
@@ -1,20 +1,21 @@
...
-#include "point.h"
 #include <array>
 #include <fstream>
 #include <iostream>
 #include <limits>
 #include <map>
 #include <memory>
 #include <vector>
-
 // libraries
```
Comment (kks32) [inline include/gmsh.h]: You don't need to explicity say local or libraries.

File: include/gmsh.h [hunk 13]
```diff
@@ -1,20 +1,21 @@
...
 #include <Eigen/Dense>
+// local
+#include "mesh.h"
+#include "point.h"
 
 //! \brief Generate Material Points from GMSH file
+template <unsigned Tdim, unsigned Tvert, unsigned Tncoords>
```
Comment (kks32) [inline include/gmsh.h]: You don't need `Tncoords` use the product of Tdim * Tvertices.

Comment (kks32) [inline include/gmsh.h]: Rename `Tvert` as `Tvertices`

File: include/gmsh.h [hunk 14]
```diff
@@ -37,38 +38,40 @@ class GMSH : public Mesh {
...
+  std::map<double, std::array<unsigned, Tvert>> elements() const {
+    return elements_;
+  }
 
   //! Return a vector of stresses
   std::vector<std::array<double, 6>> stress() const { return stress_; }
 
  private:
-  //! Number of vertices
+  //! Number of vertices in total
```
Comment (kks32) [inline include/gmsh.h]: //! Total number of vertices

File: include/gmsh.tcc [hunk 15]
```diff
@@ -63,14 +65,11 @@ void GMSH::read_elements(const std::string& filename) {
...
   const unsigned toplines = 4;
-  const unsigned ndimension = 3;
   // specify element type 4 = tetrahedral
   const unsigned element_type = 4;
-  // specify element vertices 4 = tetrahedral
-  const unsigned elementvertices = 4;
 
   //! Array to store vertices coordinates
-  std::array<double, elementvertices> elementarray;
+  std::array<double, Tvert> elementarray;
```
Comment (kks32) [inline include/gmsh.tcc]: Change `double` to `int` or `unsigned` for all ids

File: include/gmsh.tcc [hunk 16]
```diff
@@ -213,7 +206,8 @@ void GMSH::compute_material_points() {
   std::cout << "Number of Material Points: " << materialpoints_.size() << '\n';
 }
 //! Compute stresses
-void GMSH::compute_stresses() {
+template <unsigned Tdim, unsigned Tvert, unsigned Tncoords>
+void GMSH<Tdim, Tvert, Tncoords>::compute_stresses() {
 
   double density = 22;
```
Comment (kks32) [inline include/gmsh.tcc]: These should be const

File: src/main.cc [hunk 17]
```diff
@@ -16,7 +16,9 @@ int main(int argc, char** argv) {
     std::unique_ptr<IO> io(new IO(argv[1]));
 
     //! Mesh handler
-    std::unique_ptr<GMSH> mesh(new GMSH());
+    //! \Param[in] Tdim = Dimension, Tvert = n vertices in element, Tcoord =
+    //! Tdim*Tvert
+    std::unique_ptr<GMSH<3, 4, 12>> mesh(new GMSH<3, 4, 12>());
```
Comment (kks32) [inline src/main.cc]: This should be a `Mesh` class object, not `GMSH`

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/gmsh.tcc (hunk 8 above)
Comment (ezrayst) [reply include/gmsh.tcc]: I will use `Tvertices` instead

File: include/gmsh.tcc [hunk 18]
```diff
@@ -1,11 +1,12 @@
 //! Read vertex id and coordinates in GMSH
 //! \param[in] filename Input GMSH file
-void GMSH::read_vertices(const std::string& filename) {
+//! \Param[in] Tdim = Dimension, Tvert = n vertices in element
```
Comment (kks32) [inline include/gmsh.tcc]: Please read the Doxygen style for variables.
`\tparam Tdim Dimension`

File: include/gmsh.tcc [hunk 19]
```diff
@@ -28,29 +29,30 @@ void GMSH::read_vertices(const std::string& filename) {
...
-  std::cout << "Number of Vertices: " << vertices_.size() << '\n';
-  nvertices_ = vertices_.size();
+  std::cout << "Number of Vertices: " << Mesh<Tdim, Tvertices>::vertices_.size()
+            << '\n';
 }
 
 //! Read GMSH elements
 //! \param[in] filename Input GMSH file
-void GMSH::read_elements(const std::string& filename) {
+//! \Param[in] Tdim = Dimension, Tvert = n vertices in element
```
Comment (kks32) [inline include/gmsh.tcc]: `\tparam`

File: include/gmsh.tcc [hunk 20]
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
         istream >> elementry;
 
         //! \brief Check element type
-        //! \details If element type not == to Tdim, skip element
-        if (elementtype != ndimension) {
+        //! \details If element type not == to specified Tvertices, skip
```
Comment (kks32) [inline include/gmsh.tcc]: No `\details` inside the function!

File: include/gmsh.tcc [hunk 21]
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
...
         }
       }
     }
     infile.close();
   }
+  std::cout << "Number of Elements: " << Mesh<Tdim, Tvertices>::elements_.size()
+            << '\n';
+
+  //! Get the coordinates for each vertex of each element
+  GMSH::store_element_vertices();
```
Comment (kks32) [inline include/gmsh.tcc]: use `this->`

File: include/gmsh.tcc [hunk 22]
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
...
+}
+
+//! \brief store each coordinates of an element
+//! \Param[in] Tdim = Dimension, Tvert = n vertices in element
```
Comment (kks32) [inline include/gmsh.tcc]: Remove `=` and should be `\tparam`

File: include/gmsh.tcc [hunk 23]
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
...
+  std::array<double, Tvertices> elementkeyvalues;
+  std::array<double, Tdim * Tvertices> verticesarray;
+
+  //! Iterate through element_
+  for (unsigned i = firstelement; i <= lastelement; ++i) {
+    elementfind = Mesh<Tdim, Tvertices>::elements_.find(i);
+    if (elementfind != Mesh<Tdim, Tvertices>::elements_.end()) {
+
+      //! In each element, iterate to get vertices id's of the element
+      for (unsigned j = 0; j < 4; ++j) {
```
Comment (kks32) [inline include/gmsh.tcc]: Can we not combine these two for loops? Shouldn't this be iterating to Tvertices?

File: include/gmsh.tcc [hunk 24]
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
...
+}
+
+//! \brief generate material points from elements
+//! \Param[in] Tdim = Dimension, Tvert = n vertices in element
+template <unsigned Tdim, unsigned Tvertices>
+void GMSH<Tdim, Tvertices>::compute_material_points() {
+
+  unsigned arrayposition = 0;
+
+  typename std::map<unsigned, std::array<double, Tdim * Tvertices>>::iterator
```
Comment (kks32) [inline include/gmsh.tcc]: Use `alias`

File: include/gmsh.tcc [hunk 25]
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
...
+      coordinatesfind;
+
+  const unsigned firstelementcoord =
+      Mesh<Tdim, Tvertices>::elementcoordinates_.begin()->first;
+  const unsigned lastelementcoord =
+      Mesh<Tdim, Tvertices>::elementcoordinates_.rbegin()->first;
+
+  for (unsigned t = firstelementcoord; t < lastelementcoord + 1; ++t) {
```
Comment (kks32) [inline include/gmsh.tcc]: Why not use an iterator?

File: include/gmsh.tcc [hunk 26]
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
...
+
+  typename std::map<unsigned, std::array<double, Tvertices>>::iterator
+      elementfind;
+  typename std::map<unsigned, std::array<double, Tdim>>::iterator verticesfind;
+
+  std::array<double, Tvertices> elementkeyvalues;
+  std::array<double, Tdim * Tvertices> verticesarray;
+
+  //! Iterate through element_
+  for (unsigned i = firstelement; i <= lastelement; ++i) {
```
Comment (kks32) [inline include/gmsh.tcc:157]: Use an iterator.

File: include/gmsh.tcc [hunk 27]
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
...
+        }
+      }
+
+      //! Initialize array to contain points (need re-initialize as 0)
+      std::array<double, Tdim> pointsarray{0, 0, 0};
 
-  std::cout << "Number of Elements: " << elements_.size() << '\n';
+      //! Centroid test
+      //++++++++++++++++++++++++++++++++++++++
+      for (unsigned i = 0; i < 3; ++i) {
```
Comment (kks32) [inline include/gmsh.tcc]: i < Tdim, j < Tvertices

File: include/gmsh.tcc [hunk 28]
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
...
 
-  double density = 22;
-  double k0 = 0.5;
-  double max_height = 3;
-  double conv_factor = 10;
+  const double density = 22;
+  const double k0 = 0.5;
+  const double max_height = 3;
+  const double conv_factor = 10;
+  const unsigned stress_comp = 6;
```
Comment (kks32) [inline include/gmsh.tcc]: This is just Tdim * 2, use Tdim * 2 instead.

File: include/gmsh.tcc [hunk 29]
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
...
     hor_stress = ver_stress * k0;
-    std::array<double, 6> stress{hor_stress, hor_stress, ver_stress, 0, 0, 0};
-    stress_.emplace_back(stress);
+    std::array<double, stress_comp> stress{hor_stress, hor_stress, ver_stress,
+                                           0,          0,          0};
+    Mesh<Tdim, Tvertices>::stress_.emplace_back(stress);
   }
 
-  std::cout << "Compute initial stresses for material points.\n";
+  std::cout << "Initial stresses for material points computed.\n";
```
Comment (kks32) [inline include/gmsh.tcc]: Remove cout

File: include/io.tcc [hunk 30]
```diff
@@ -1,35 +1,36 @@
 //! \brief Write output file for point
 //! \details Get stress vector of vertex coordinates, and number of vertices
 void IO::write_vertices(
-    const std::vector<std::shared_ptr<Point<3>>>& vertices) {
+    const std::vector<std::shared_ptr<Point<3>>>& materialpoints) {
 
   //! Output vertices file
   std::fstream vertices_file;
```
Comment (kks32) [inline include/io.tcc]: Change vertices to material points everywher.

File: include/mesh.h [hunk 31]
```diff
@@ -3,19 +3,78 @@
...
 
+#include "point.h"
+
 //! \brief Abstract class for handling mesh
+template <unsigned Tdim, unsigned Tvertices>
 class Mesh {
 
  public:
-  //! Read vertices from mesh
+  //! \brief constructor for a mesh class
```
Comment (kks32) [inline include/mesh.h]: Remove these!

File: include/mesh.h [hunk 32]
```diff
@@ -3,19 +3,78 @@
...
+  virtual void compute_material_points() = 0;
+
+  //! Compute stresses
   virtual void compute_stresses() = 0;
+
+  //! call total number of vertices generated
+  virtual unsigned nvertices() const { return 0; }
+
+  //! Return a vector of material points
+  virtual std::vector<std::shared_ptr<Point<Tdim>>> material_points() {
```
Comment (kks32) [inline include/mesh.h]: Why are these virtual?

File: include/mesh.h [hunk 33]
```diff
@@ -3,19 +3,78 @@
...
+    return std::vector<std::shared_ptr<Point<Tdim>>>{};
+  }
+
+  //! Return a map of mesh element vertices
+  virtual std::map<unsigned, std::array<double, Tdim>> vertices() const {
```
Comment (kks32) [inline include/mesh.h]: Shouldn't be virtual

File: include/mesh.h [hunk 34]
```diff
@@ -3,19 +3,78 @@
...
+    return std::map<unsigned, std::array<double, Tdim>>{};
+  }
+
+  //! Return a map of element id & vertices id
+  virtual std::map<unsigned, std::array<double, Tvertices>> elements() const {
```
Comment (kks32) [inline include/mesh.h]: Not virutal

File: include/mesh.h [hunk 35]
```diff
@@ -3,19 +3,78 @@
...
+    return std::map<unsigned, std::array<double, Tvertices>>{};
+  }
+
+  //! Return a vector of stresses
+  virtual std::vector<std::array<double, 6>> stress() const {
```
Comment (kks32) [inline include/mesh.h]: Not virutal

File: include/mesh.h [hunk 36]
```diff
@@ -3,19 +3,78 @@
...
+    return std::vector<std::array<double, 6>>{};
+  }
+
+ protected:
+  //! Total number of vertices
+  unsigned nvertices_;
+
+  std::vector<std::shared_ptr<Point<Tdim>>> materialpoints_;
```
Comment (kks32) [inline include/mesh.h]: Add comment

File: include/mesh.h [hunk 37]
```diff
@@ -3,19 +3,78 @@
...
+
+  //! Stress vector in Voigt Notation
+  //! $\sigma_{xx}$ $\sigma_{yy}$ $\sigma_{zz}$
+  //! $\tau_{yz}$ $\tau_{zx}$ $\tau_{xy}$
+  std::vector<std::array<double, 6>> stress_;
```
Comment (kks32) [inline include/mesh.h]: Should be Tdim * 2

File: include/gmsh.h [hunk 38]
```diff
@@ -21,34 +25,37 @@ class GMSH : public Mesh {
...
+  }
 
-  //! Vector of elements
-  std::vector<std::shared_ptr<Point<3>>> elements_;
-
-  //! Stress vector in Voigt Notation
-  //! $\sigma_{xx}$ $\sigma_{yy}$ $\sigma_{zz}$ $\tau_{yz}$ $\tau_{zx}$ $\tau_{xy}$
-  std::vector<std::array<double, 6>> stress_;
+  //! Return a vector of stresses
+  std::vector<std::array<double, 6>> stress() const {
```
Comment (kks32) [inline include/gmsh.h]: Should be defined in `Mesh`

File: include/gmsh.h [hunk 39]
```diff
@@ -21,34 +25,37 @@ class GMSH : public Mesh {
...
-  unsigned nvertices_;
+  //! Return a map of mesh element vertices
+  std::map<unsigned, std::array<double, Tdim>> vertices() const {
+    return Mesh<Tdim, Tvertices>::vertices_;
+  }
 
-  //! Vector of vertices 
-  std::vector<std::shared_ptr<Point<3>>> vertices_;
+  //! Return a map of element id & vertices id
+  std::map<unsigned, std::array<double, Tvertices>> elements() const {
```
Comment (kks32) [inline include/gmsh.h]: Should be defined in `Mesh`

File: include/gmsh.h [hunk 40]
```diff
@@ -21,34 +25,37 @@ class GMSH : public Mesh {
...
+  }
 
-  //! Return a vector of stresses
-  std::vector<std::array<double, 6>> stress() const { return stress_; }
-
- private:
-  //! Number of vertices
-  unsigned nvertices_;
+  //! Return a map of mesh element vertices
+  std::map<unsigned, std::array<double, Tdim>> vertices() const {
```
Comment (kks32) [inline include/gmsh.h]: Should be defined in `Mesh`

File: include/gmsh.h [hunk 41]
```diff
@@ -21,34 +25,37 @@ class GMSH : public Mesh {
...
-  unsigned nvertices() const { return nvertices_; }
-
-  //! Return a vector of vertices
-  std::vector<std::shared_ptr<Point<3>>> vertices() const { return vertices_; }
+  unsigned nvertices() const { return Mesh<Tdim, Tvertices>::nvertices_; }
 
-  //! Return a vector of elements
-  std::vector<std::shared_ptr<Point<3>>> elements() const { return elements_; }
+  //! Return a vector of material points
+  std::vector<std::shared_ptr<Point<Tdim>>> material_points() {
```
Comment (kks32) [inline include/gmsh.h]: Should be defined in Mesh

File: include/gmsh.h [hunk 42]
```diff
@@ -21,34 +25,37 @@ class GMSH : public Mesh {
...
+
   //! Compute stresses
   void compute_stresses();
 
   //! call total number of vertices generated
-  unsigned nvertices() const { return nvertices_; }
-
-  //! Return a vector of vertices
-  std::vector<std::shared_ptr<Point<3>>> vertices() const { return vertices_; }
+  unsigned nvertices() const { return Mesh<Tdim, Tvertices>::nvertices_; }
```
Comment (kks32) [inline include/gmsh.h]: Should be defined in Mesh

File: include/gmsh.h (hunk 38 above)
Comment (kks32) [reply include/gmsh.h]: Use the Mesh class's `stress_`.  using `Mesh<Tdim, Tvertices>::stress_`

File: src/main.cc [hunk 43]
```diff
@@ -16,14 +17,18 @@ int main(int argc, char** argv) {
     std::unique_ptr<IO> io(new IO(argv[1]));
 
     //! Mesh handler
-    std::unique_ptr<GMSH> mesh(new GMSH());
+    //! \Param[in] Tdim = Dimension, Tvert = n vertices in element
```
Comment (kks32) [inline src/main.cc]: Why `\Param` in the middle of the main function?

File: src/main.cc [hunk 44]
```diff
@@ -16,14 +17,18 @@ int main(int argc, char** argv) {
...
+    GMSH<3, 4> gmsh;
```
Comment (kks32) [inline src/main.cc]: GMSH should not appear!

File: src/main.cc [hunk 45]
```diff
@@ -16,14 +17,18 @@ int main(int argc, char** argv) {
...
+    Mesh<3, 4>* mesh = &gmsh;
```
Comment (kks32) [inline src/main.cc]: Use a unique_ptr of base class `Mesh` but create it with a new `GMSH`

File: src/main.cc [hunk 46]
```diff
@@ -16,14 +17,18 @@ int main(int argc, char** argv) {
...
+    //! Read mesh and compute material points & stresses
     mesh->read_vertices(io->mesh_file_name());
+    mesh->read_elements(io->mesh_file_name());
+    mesh->compute_material_points();
     mesh->compute_stresses();
 
-    //! Write vertices and stresses
-    io->write_vertices(mesh->vertices());
+    //! Write material points and stresses
+    io->write_vertices(mesh->material_points());
```
Comment (kks32) [inline src/main.cc]: should be `write_material_points`

**Review by kks32 — CHANGES_REQUESTED** [review]
Check inheritance and base class / derived class of Mesh and GMSH.

File: include/gmsh.tcc [hunk 47]
```diff
@@ -28,29 +29,30 @@ void GMSH::read_vertices(const std::string& filename) {
...
         //! Coordinates of vertex
-        std::array<double, ndimension> vertex;
-        
+        std::array<double, Tdim> vertex;
+
         istream >> vertid;
         istream >> vertex.at(0) >> vertex.at(1) >> vertex.at(2);
-
-        vertices_.emplace_back(new Point<ndimension>(vertid, vertex));
+        Mesh<Tdim, Tvertices>::vertices_.insert(std::make_pair(vertid, vertex));
```
Comment (kks32) [inline include/gmsh.tcc]: Use `using Mesh<Tdim, Tvertices>::vertices_` in the header file, and you can use `vertices_`.

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/gmsh.tcc [hunk 48]
```diff
@@ -1,6 +1,6 @@
 //! Read vertex id and coordinates in GMSH
+//! \tparam Tdim Dimension, Tvert Number of vertices in element
```
Comment (kks32) [inline include/gmsh.tcc]: `Tvert` should be in the next line. Use `doxygen` to check if the comments are generated properly.

File: include/gmsh.tcc [hunk 49]
```diff
@@ -49,8 +49,8 @@ void GMSH<Tdim, Tvertices>::read_vertices(const std::string& filename) {
 }
 
 //! Read GMSH elements
+//! \tparam Tdim Dimension, Tvert Number of vertices in element
```
Comment (kks32) [inline include/gmsh.tcc]: Same `Tvert` in a new line

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/gmsh.h [hunk 50]
```diff
@@ -21,34 +23,46 @@ class GMSH : public Mesh {
...
+
   //! Compute stresses
   void compute_stresses();
 
   //! call total number of vertices generated
   unsigned nvertices() const { return nvertices_; }
 
-  //! Return a vector of vertices
-  std::vector<std::shared_ptr<Point<3>>> vertices() const { return vertices_; }
+  //! Return a map of mesh element vertices
```
Comment (kks32) [inline include/gmsh.h]: This should be in the `Mesh` class

File: include/gmsh.h [hunk 51]
```diff
@@ -1,18 +1,20 @@
...
 #include <memory>
 #include <vector>
 
-#include "mesh.h"
-#include "point.h"
-
 //! \brief Generate Material Points from GMSH file
-class GMSH : public Mesh {
+template <unsigned Tdim, unsigned Tvertices>
+class GMSH : public Mesh<Tdim> {
```
Comment (kks32) [inline include/gmsh.h]: Templatise the mesh with `Tvertices` as well.

File: include/gmsh.h [hunk 52]
```diff
@@ -21,34 +23,46 @@ class GMSH : public Mesh {
...
-  std::vector<std::shared_ptr<Point<3>>> vertices() const { return vertices_; }
+  //! Return a map of mesh element vertices
+  std::map<unsigned , std::array<unsigned, Tdim>> vertices() const {
+    return vertices_;
+  }
 
-  //! Return a vector of elements
-  std::vector<std::shared_ptr<Point<3>>> elements() const { return elements_; }
+  //! Return a map of element id & vertices id
+  std::map<unsigned, std::array<unsigned, Tvertices>> elements() const {
```
Comment (kks32) [inline include/gmsh.h]: This should go to the `Mesh` class

File: include/gmsh.h [hunk 53]
```diff
@@ -21,34 +23,46 @@ class GMSH : public Mesh {
...
+    return elements_;
+  }
 
   //! Return a vector of stresses
-  std::vector<std::array<double, 6>> stress() const { return stress_; }
+  std::vector<std::array<double, Tdim * 2>> stress() const { return stress_; }
 
  private:
-  //! Number of vertices
+  //! Number of vertices in total
```
Comment (kks32) [inline include/gmsh.h]: `//! Total number of vertices`

File: include/gmsh.h [hunk 54]
```diff
@@ -21,34 +23,46 @@ class GMSH : public Mesh {
...
   unsigned nvertices_;
 
-  //! Vector of vertices 
-  std::vector<std::shared_ptr<Point<3>>> vertices_;
+  //! Stress vector in Voigt Notation
```
Comment (kks32) [inline include/gmsh.h]: `stress_`, `vertices_`, and `elementcoords_` should be in base class `Mesh`.

File: include/gmsh.h [hunk 55]
```diff
@@ -21,34 +23,46 @@ class GMSH : public Mesh {
...
-  std::vector<std::shared_ptr<Point<3>>> elements_;
+  //! Map to store id and vertices coordinates
+  std::map<double, std::array<double, Tdim>> vertices_;
 
-  //! Stress vector in Voigt Notation
-  //! $\sigma_{xx}$ $\sigma_{yy}$ $\sigma_{zz}$ $\tau_{yz}$ $\tau_{zx}$ $\tau_{xy}$
-  std::vector<std::array<double, 6>> stress_;
+  //! Map to store element ID and vertices ID
+  std::map<double, std::array<double, Tvertices>> elements_;
+  std::map<double, std::array<double, Tdim * Tvertices>> elementcoordinates_;
```
Comment (kks32) [inline include/gmsh.h]: Why coordinates of elements, is it not sufficient to map it to the ID of vertices?

File: include/gmsh.tcc [hunk 56]
```diff
@@ -108,43 +109,130 @@ void GMSH::read_elements(const std::string& filename) {
...
+  std::array<double, Tdim * Tvertices> verticesarray;
+
+  //! Iterate through element_
+  for (unsigned q = firstelement; q <= lastelement; ++q) {
+    elementfind = elements_.find(q);
+    if (elementfind != elements_.end()) {
+
+      //! In each element, iterate to get vertices id's of the element
+      for (unsigned j = 0; j < Tvertices; ++j) {
+        elementkeyvalues[j] = elementfind->second[j];
```
Comment (kks32) [inline include/gmsh.tcc]: Use `.at` operator

File: include/gmsh.tcc [hunk 57]
```diff
@@ -108,43 +109,130 @@ void GMSH::read_elements(const std::string& filename) {
...
+      }
+      //! Iterate through the vertices to get coordinates (4 for tetrahedral)
+      for (unsigned q = 0; q < Tvertices; ++q) {
```
Comment (kks32) [inline include/gmsh.tcc]: Which `q`, defined in line 146 or 155?

File: include/gmsh.tcc [hunk 58]
```diff
@@ -108,43 +109,130 @@ void GMSH::read_elements(const std::string& filename) {
...
+          arrayposition = (i * Tdim) + j;
+          m(j, i) = coordinatesfind->second[arrayposition];
+        }
+      }
+
+      // Centroid test
+      //++++++++++++++++++++++++++++++++++++++
+      for (unsigned i = 0; i < Tdim; ++i) {
+        // reset array values
+        pointsarray[i] = 0;
```
Comment (kks32) [inline include/gmsh.tcc]: Use STL algorithms or fill

Comment (kks32) [reply include/gmsh.tcc]: Why not declare pointsarray here?

File: include/gmsh.tcc [hunk 59]
```diff
@@ -108,43 +109,130 @@ void GMSH::read_elements(const std::string& filename) {
...
+  const double k0 = 0.5;
+  const double max_height = 3;
+  const double conv_factor = 10;
   double ver_stress;
   double hor_stress;
 
   //! Loop through the points to get vertical and horizontal stresses
   //! Note that tau (shear stress) is assumed 0
-  for (const auto& point : vertices_) {
+  for (const auto& point : Mesh<Tdim>::materialpoints_) {
```
Comment (kks32) [inline include/gmsh.tcc]: Use `using` in the derived class to avoid these `Mesh<Tdim>::`

File: include/io.tcc [hunk 60]
```diff
@@ -40,14 +46,17 @@ void IO::write_stresses(const std::vector<std::array<double, 6>>& stresses) {
     stress_file << stresses.size() << "\n";
 
     //! Stresses in Voigt Notation
+    // Stress needs ID too
```
Comment (kks32) [inline include/io.tcc]: OK!

File: include/mesh.h [hunk 61]
```diff
@@ -1,21 +1,47 @@
...
+  virtual void read_elements(const std::string& filename) = 0;
+
+  //!
+  virtual void compute_material_points() = 0;
+
   //! Compute initial stresses for material points
   virtual void compute_stresses() = 0;
+
+  //! Return a vector of material points
+  std::vector<std::shared_ptr<Point<Tdim>>> write_material_points() {
```
Comment (kks32) [inline include/mesh.h]: rename to `material_points` as this function doesn't write anything.

File: include/mesh.h [hunk 62]
```diff
@@ -1,21 +1,47 @@
...
+    return materialpoints_;
+  }
+
+  //! Return a vector of stresses
+  std::vector<std::array<double, Tdim * 2>> write_stresses() const {
+    return stress_;
+  }
+
+  //! Container for storing material points
```
Comment (kks32) [inline include/mesh.h]: Shouldn't these be in `protected`?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/gmsh.h [hunk 63]
```diff
@@ -14,7 +14,7 @@
```
Comment (kks32) [inline include/gmsh.h:11]: Could you run `clang-format`, the includes are not in the right order.

**Review by kks32 — CHANGES_REQUESTED** [review]

File: src/main.cc [hunk 64]
```diff
@@ -18,7 +18,7 @@ int main(int argc, char** argv) {
     //! Mesh handler
     //! \Param[in] Tdim = Dimension, Tvert = n vertices in element
 
-    std::unique_ptr<Mesh<3,4>> mesh(new GMSH<3, 4>());
+    std::unique_ptr<Mesh<3, 4>> mesh(new GMSH<3, 4>());
 
     //! Read mesh and compute material points & stresses
     mesh->read_vertices(io->mesh_file_name());
```
Comment (kks32) [inline src/main.cc]: Could we have a single function? `read_mesh` which calls both the vertices and elements?

**Review by kks32 — CHANGES_REQUESTED** [review]

**Review by kks32 — APPROVED** [review]

**Pull request merged**